### ISSUE TITLE
Can't execute code on Google Colab

### ISSUE THREAD

**Issue by arthurdeschamps** [locuslab/SATNet#6]
Hi!
I've been trying to run the code provided on Google Colab but it produces an error when running the code cell under "**The 9x9 Sudoku Experiment**". The error occurs in PyTorch, and I've tried downgrading to versions < 1.3.1 but there is another dependency that requires one to use version 1.3.1.

Here is the full stack trace:

`---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-13-8c6721fcbe76> in <module>()
     10 test_logger = FigLogger(fig, axes[1], 'Testing')
     11 
---> 12 test(args.boardSz, 0, sudoku_model, optimizer, test_logger, sudoku_test, args.batchSz)
     13 plt.pause(0.01)
     14 for epoch in range(1, args.nEpoch+1):

6 frames
/usr/local/lib/python3.6/dist-packages/torch/autograd/grad_mode.py in decorate_no_grad(*args, **kwargs)
     47         def decorate_no_grad(*args, **kwargs):
     48             with self:
---> 49                 return func(*args, **kwargs)
     50         return decorate_no_grad
     51 

/content/SATNet/exps/sudoku.py in test(args, epoch, model, optimizer, logger, dataset, batchSz, unperm)
    270 @torch.no_grad()
    271 def test(args, epoch, model, optimizer, logger, dataset, batchSz, unperm=None):
--> 272     run(args, epoch, model, optimizer, logger, dataset, batchSz, False, unperm)
    273 
    274 @torch.no_grad()

/content/SATNet/exps/sudoku.py in run(boardSz, epoch, model, optimizer, logger, dataset, batchSz, to_train, unperm)
    251             optimizer.step()
    252 
--> 253         err = computeErr(preds.data, boardSz, unperm)/batchSz
    254         tloader.set_description('Epoch {} {} Loss {:.4f} Err: {:.4f}'.format(epoch, ('Train' if to_train else 'Test '), loss.item(), err))
    255         loss_final += loss.item()

/usr/local/lib/python3.6/dist-packages/torch/autograd/grad_mode.py in decorate_no_grad(*args, **kwargs)
     47         def decorate_no_grad(*args, **kwargs):
     48             with self:
---> 49                 return func(*args, **kwargs)
     50         return decorate_no_grad
     51 

/content/SATNet/exps/sudoku.py in computeErr(pred_flat, n, unperm)
    292     for j in range(nsq):
    293         # Check the jth row and column.
--> 294         boardCorrect[invalidGroups(I[:,j,:])] = 0
    295         boardCorrect[invalidGroups(I[:,:,j])] = 0
    296 

/content/SATNet/exps/sudoku.py in invalidGroups(x)
    287         valid *= (x.max(1)[0] == nsq-1)
    288         valid *= (x.sum(1) == s)
--> 289         return 1-valid
    290 
    291     boardCorrect = torch.ones(batchSz).type_as(pred)

/usr/local/lib/python3.6/dist-packages/torch/tensor.py in __rsub__(self, other)
    344 
    345     def __rsub__(self, other):
--> 346         return _C._VariableFunctions.rsub(self, other)
    347 
    348     def __rdiv__(self, other):

RuntimeError: Subtraction, the `-` operator, with a bool tensor is not supported. If you are trying to invert a mask, use the `~` or `logical_not()` operator instead.`

**Comment by rocknamx8** [comment]
I also meet the same problem :(

**Comment by xflash96** [comment]
PyTorch 1.2 introduced the bool type to replace the uint8 and it is not backward compatible. The following diff should fix this bug in PyTorch version >= 1.2. 

```diff
diff --git a/exps/sudoku.py b/exps/sudoku.py
index 657d11a..162bf40 100644
--- a/exps/sudoku.py
+++ b/exps/sudoku.py
@@ -286,7 +286,7 @@ def computeErr(pred_flat, n, unperm):
         valid = (x.min(1)[0] == 0)
         valid *= (x.max(1)[0] == nsq-1)
         valid *= (x.sum(1) == s)
-        return 1-valid
+        return valid.bitwise_not()

     boardCorrect = torch.ones(batchSz).type_as(pred)
     for j in range(nsq):
```

**Comment by xflash96** [comment]
Sorry for the delay. I've matched the API with Pytorch:1.7.0, and now the example works on Colab.
Am closing the issue.
